File: chanceapi-java/src/main/java/ru/dverkask/chance/PercentageChanceTable.java
package ru.dverkask.chance;

public class PercentageChanceTable<T> extends ChanceTable<T> {
    @SafeVarargs protected PercentageChanceTable(Chance<T>... chances) {
        for (Chance<T> chance : chances) {
            this.addChance(chance.item(), chance.chance());
        }
    }
    @Override
    public void addChance(T item, double chance) {
        if (chance < 0 || chance > 1) {
            throw new IllegalArgumentException("Chance must be between 0 and 1");
        }
        chances.add(new Chance<>(item, chance));
    }

    @Override
    public T roll() {
        double p = Math.random();
        double cumulativeProbability = 0.0;
        for (Chance<T> chance : chances) {
            cumulativeProbability += chance.chance();
            if (p <= cumulativeProbability) {
                return chance.item();
            }
        }
        return null;
    }
}

File: chanceapi-kotlin/src/main/kotlin/ru/dverkask/chance/PercentageChanceTable.kt
package ru.dverkask.chance

class PercentageChanceTable<T>(chances: List<Chance<T>>) : ChanceTable<T>() {
    init {
        chances.forEach { addChance(it.item, it.chance) }
    }

    override fun addChance(item: T, chance: Double) {
        require(chance in 0.0..1.0) { "Chance must be between 0 and 1" }
        chances.add(Chance(item, chance))
    }

    override fun roll(): T? {
        val p = Math.random()
        var cumulativeProbability = 0.0
        for (chance in chances) {
            cumulativeProbability += chance.chance
            if (p <= cumulativeProbability) {
                return chance.item
            }
        }
        return null
    }
}